### PR TITLE
Fix the wwwroot `Content` and `Script` folder path construction

### PR DESCRIPTION
**Pull request by ShanshanHe**
## Related issue

Closes: #84

## Description
Fix the wwwroot `Content` and `Script` folder path construction.
    
When `FolderUpdate` action was executed for MVC type of project, cta creates `wwwroot` folder if not already exists, and move `Content` an `Script` folders in it.
    
Currently the `Content` and `Script` folder paths are constructed as follows:
`<ProjectPath>\wwwrootContent`
`<ProjectPath>\wwwrootScript`
    
Instead they should be like following:
`<ProjectPath>\wwwroot\Content`
`<ProjectPath>\wwwroot\Script`
    
This patch includes the fix for the path construction.

## Supplemental testing
Unit tests have been added to verify the `Content` and `Script` folder path are correctly constructed for both `Mvc` and `WebApi` project types.

---
*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

**Review by jonlouie — CHANGES_REQUESTED** [review]
Couple of small change requests and nits, otherwise looks great

File: src/CTA.Rules.Actions/ActionHelpers/TemplateHelper.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ internal static string GetTemplateFileContent(string projectNamespace, ProjectTy
         /// <returns></returns>
         private static string GetProjectTypeTemplateDir(ProjectType projectType)
         {
-            return string.Concat("Templates", Path.DirectorySeparatorChar, projectType.ToString());
+            return string.Concat("Templates", Path.DirectorySeparatorChar, projectType.ToString().ToLower());
```
[the remark this answers is not in the document]
Comment (marknfawaz) [reply src/CTA.Rules.Actions/ActionHelpers/TemplateHelper.cs:32]: +1 to Jon's comment above. Also, what is the purpose of adding ToLower()?

**Review by marknfawaz — CHANGES_REQUESTED** [review]
Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.

**Comment by jonlouie** [comment]
> Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.

@marknfawaz The general practice is to isolate unit tests so they don't rely on any external services/resources. By using the templates in S3, the unit tests could be broken if a template in S3 is updated. 

One idea to bridge the gap between what we currently have implemented and standard practice is to include the templates with source control and sync them to S3 (similar to what's done in porting-assistant-dotnet-datastore). This would allow us to post hot-fixes to prod while decoupling tests from S3.

**Comment by marknfawaz** [comment]
> > Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.
> 
> @marknfawaz The general practice is to isolate unit tests so they don't rely on any external services/resources. By using the templates in S3, the unit tests could be broken if a template in S3 is updated.
> 
> One idea to bridge the gap between what we currently have implemented and standard practice is to include the templates with source control and sync them to S3 (similar to what's done in porting-assistant-dotnet-datastore). This would allow us to post hot-fixes to prod while decoupling tests from S3.

I agree that the general practice for unit tests is for them to be independent. However, our tests are a mixture of integration and unit tests (downloaded projects, rule files, etc...). One future enhancement would be to separate the two types of tests based on their type. My concern here is when the issues are caused by a change to these templates, how would we detect them if our tests have their own templates?

We need to add the templates to source control regardless of this use case. Issue is [here](https://github.com/aws/porting-assistant-dotnet-datastore/issues/45). Although I'm not sure how it would be different except that we're downloading from a different location.

**Comment by jonlouie** [comment]
> > > Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.
> > 
> > 
> > @marknfawaz The general practice is to isolate unit tests so they don't rely on any external services/resources. By using the templates in S3, the unit tests could be broken if a template in S3 is updated.
> > One idea to bridge the gap between what we currently have implemented and standard practice is to include the templates with source control and sync them to S3 (similar to what's done in porting-assistant-dotnet-datastore). This would allow us to post hot-fixes to prod while decoupling tests from S3.
> 
> I agree that the general practice for unit tests is for them to be independent. However, our tests are a mixture of integration and unit tests (downloaded projects, rule files, etc...). One future enhancement would be to separate the two types of tests based on their type. My concern here is when the issues are caused by a change to these templates, how would we detect them if our tests have their own templates?
> 
> We need to add the templates to source control regardless of this use case. Issue is [here](https://github.com/aws/porting-assistant-dotnet-datastore/issues/45). Although I'm not sure how it would be different except that we're downloading from a different location.

Good point about the tests- I'll create an issue to categorize unit vs integration tests.

I agree and share the same concern- test files and prod files should be identical to prevent any undetected issues. My thought was to add templates to the CTA repo (maybe in their own directory) and, if possible, sync those files to S3 upon merge. These local templates would be used for testing, and the S3 templates could be used in prod. This approach would: 
1. Resolve the templates in source control issue 
2. Allow changes to be pushed out to prod from source, rather than pulled in from prod for tests

Is it possible to sync files to S3 without going through porting-assistant-dotnet-datastore? If we can keep the templates within the CTA repo would make the dev workflow simpler and the codebase more navigable.

**Comment by Eruanion** [comment]
> 
> 
> > > > Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.
> > > 
> > > 
> > > @marknfawaz The general practice is to isolate unit tests so they don't rely on any external services/resources. By using the templates in S3, the unit tests could be broken if a template in S3 is updated.
> > > One idea to bridge the gap between what we currently have implemented and standard practice is to include the templates with source control and sync them to S3 (similar to what's done in porting-assistant-dotnet-datastore). This would allow us to post hot-fixes to prod while decoupling tests from S3.
> > 
> > 
> > I agree that the general practice for unit tests is for them to be independent. However, our tests are a mixture of integration and unit tests (downloaded projects, rule files, etc...). One future enhancement would be to separate the two types of tests based on their type. My concern here is when the issues are caused by a change to these templates, how would we detect them if our tests have their own templates?
> > We need to add the templates to source control regardless of this use case. Issue is [here](https://github.com/aws/porting-assistant-dotnet-datastore/issues/45). Although I'm not sure how it would be different except that we're downloading from a different location.
> 
> Good point about the tests- I'll create an issue to categorize unit vs integration tests.
> 
> I agree and share the same concern- test files and prod files should be identical to prevent any undetected issues. My thought was to add templates to the CTA repo (maybe in their own directory) and, if possible, sync those files to S3 upon merge. These local templates would be used for testing, and the S3 templates could be used in prod. This approach would:
> 
>     1. Resolve the templates in source control issue
> 
>     2. Allow changes to be pushed out to prod from source, rather than pulled in from prod for tests
> 
> 
> Is it possible to sync files to S3 without going through porting-assistant-dotnet-datastore? If we can keep the templates within the CTA repo would make the dev workflow simpler and the codebase more navigable.

I second Jon's opinion above regarding the need for these templates to be in source control but they do appear to belong logically with the datatore repo, since these templates are ultimately rules on a bigger scale however and should be part of the same repo as customer could ultimately contribute on these templates eventually.

I do agree however that downloading from S3 for unit tests is not ideal. We could require the datastore repo be installed locallly to run the unit tests locally and grab the files locally to avoid having to go to prod to fetch them. As a backup if the repo is not present on the local machine we could fall back on S3 bucket as a last resort.

File: src/CTA.Rules.Actions/ActionHelpers/TemplateHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShanshanHe) [reply src/CTA.Rules.Actions/ActionHelpers/TemplateHelper.cs:32]: Thank you for the comment, agree, I will add a constant for it and change to `Path.Combine`.

To mark's question regarding `ToLower`, all the sub folders under `Templates` are having lower case names, however, the `projectType.Tostring` returns the project type with first letter being capital letter (e.g. `Mvc`, `ClassLibrary` etc). For windows system, file path/name are not case sensitive, for for linux system it is, changing it to lower case will ensure it works (as well as the unit tests) on linux too.

File: tst/CTA.Rules.Test/FolderUpdateTests.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        private void CreateEmptyXmlDocument(string filePath)
+        {
+            File.WriteAllText(filePath, "<root></root>");
+        }
+
+        [Test]
+        public void Folder_Update_for_Mvc_Project()
+        {
+            // Create MVC Content and Script foldel under MVC test project folderr
```
[the remark this answers is not in the document]
Comment (ShanshanHe) [reply tst/CTA.Rules.Test/FolderUpdateTests.cs:41]: nice catch, will update it.

**Comment by ShanshanHe** [comment]
> > > > > Looks good overall. I added One small question in addition to the other comments. Also, is there a reason we are adding the templates to the test project? The template files can be downloaded from S3 instead.
> > > > 
> > > > 
> > > > @marknfawaz The general practice is to isolate unit tests so they don't rely on any external services/resources. By using the templates in S3, the unit tests could be broken if a template in S3 is updated.
> > > > One idea to bridge the gap between what we currently have implemented and standard practice is to include the templates with source control and sync them to S3 (similar to what's done in porting-assistant-dotnet-datastore). This would allow us to post hot-fixes to prod while decoupling tests from S3.
> > > 
> > > 
> > > I agree that the general practice for unit tests is for them to be independent. However, our tests are a mixture of integration and unit tests (downloaded projects, rule files, etc...). One future enhancement would be to separate the two types of tests based on their type. My concern here is when the issues are caused by a change to these templates, how would we detect them if our tests have their own templates?
> > > We need to add the templates to source control regardless of this use case. Issue is [here](https://github.com/aws/porting-assistant-dotnet-datastore/issues/45). Although I'm not sure how it would be different except that we're downloading from a different location.
> > 
> > 
> > Good point about the tests- I'll create an issue to categorize unit vs integration tests.
> > I agree and share the same concern- test files and prod files should be identical to prevent any undetected issues. My thought was to add templates to the CTA repo (maybe in their own directory) and, if possible, sync those files to S3 upon merge. These local templates would be used for testing, and the S3 templates could be used in prod. This approach would:
> > ```
> > 1. Resolve the templates in source control issue
> > 
> > 2. Allow changes to be pushed out to prod from source, rather than pulled in from prod for tests
> > ```
> > 
> > 
> > Is it possible to sync files to S3 without going through porting-assistant-dotnet-datastore? If we can keep the templates within the CTA repo would make the dev workflow simpler and the codebase more navigable.
> 
> I second Jon's opinion above regarding the need for these templates to be in source control but they do appear to belong logically with the datatore repo, since these templates are ultimately rules on a bigger scale however and should be part of the same repo as customer could ultimately contribute on these templates eventually.
> 
> I do agree however that downloading from S3 for unit tests is not ideal. We could require the datastore repo be installed locallly to run the unit tests locally and grab the files locally to avoid having to go to prod to fetch them. As a backup if the repo is not present on the local machine we could fall back on S3 bucket as a last resort.

Great point about separating unit test and integration test, thank you @jonlouie for creating the ticket.

I also think it's a good idea to source/version control it, it's more maintainable and can avoid version discrepancy between cta and the template files . Out of curiosity, would it be necessary to sync it to S3 and have CTA download it if the templates are already included in the project? Because I thought they can be directly consumed if they are packaged with CTA.

**Review by jonlouie — DISMISSED** [review]

**Review by Eruanion — APPROVED** [review]

**Review by jonlouie — COMMENTED** [review]
LGTM, but if we do want to remove the template files, we can shorten the scope of the unit tests to just test the private `CreateMvcDirs()` method. This approach would avoid the use of any of the template files so we could then delete them safely.

Here is one example of testing a private method: 
https://github.com/aws/cta/blob/150656c07bcd7fbd5a2439a538901aadda1b0a4c/tst/CTA.FeatureDetection.Tests/CTA.FeatureDetection.AuthType/AuthTypeFeatureTests.cs#L68-L77

**Comment by ShanshanHe** [comment]
> LGTM, but if we do want to remove the template files, we can shorten the scope of the unit tests to just test the private `CreateMvcDirs()` method. This approach would avoid the use of any of the template files so we could then delete them safely.
> 
> Here is one example of testing a private method:
> 
> https://github.com/aws/cta/blob/150656c07bcd7fbd5a2439a538901aadda1b0a4c/tst/CTA.FeatureDetection.Tests/CTA.FeatureDetection.AuthType/AuthTypeFeatureTests.cs#L68-L77

Thanks for sharing the example on testing private method @jonlouie, this is really good to know. The only concern I have by changing the test to only test the private `CreateMvcDirs` is that the scope of the tests will be reduced. Currently the test covers the [following code path:](https://github.com/aws/cta/pull/105/files#diff-ac5b8a50ec3151b2ebc6d47109a6989c5e0547996c6e962b5b407f8ad8e1d82dL34), by changing it, we won't be able to cover the two `if` paths.

**Comment by ShanshanHe** [comment]
Per our discussion, we will keep the current test cases. As for `Templates` files needed for the tests, instead of including/versioning them in the test project, we will change the test behavior to download them from S3 bucket during the test setup.

**Review by marknfawaz — APPROVED** [review]

**Review by Eruanion — APPROVED** [review]

**Review by jonlouie — APPROVED** [review]

**Pull request merged**